src/PureScript/Backend/Optimizer/Semantics.purs

```diff
@@ -621,6 +627,16 @@ evalPrimOp env = case _ of
           NeutPrimOp (Op1 op1 x')
   Op2 op2 x y ->
     case op2 of
+      OpArrayIndex
+        | NeutLit (LitArray arr) <- deref x
+        , NeutLit (LitInt i) <- deref y
+        , neut <- unsafePartial $ Array.unsafeIndex arr i
+        , Just _ <-
+            oneOf
+              [ void $ caseInt neut
+              , void $ caseNumber neut
+              , void $ caseString neut
+              ] -> neut
```

Comment:
I don't think this optimization should happen here. In the case you are testing, you are left with code that only indexes into a known array. I would expect the optimizer to unpack the array in that case. We don't have an UnpackArray rewrite, but we should.

Reply:
I would recommend backing out this, and the associated test. And we can look at adding an unpack array rewrite in another PR.